rfcs/20190722-tflite-training.md

```diff
@@ -0,0 +1,444 @@
+    has a tool which feeds the same random data into TensorFlow and TensorFlow
+    Lite, and compare if the results are close enough in a threshold. The model
+    should be extended to handle multiple subgraphs (including training and
+    inference subgraphs) and variables. This is an end to end test which can
+    capture kernel bugs or TensorFlow/TensorFlow Lite discrepancy.
+
+## Implementation details for basic training
+
+There is nothing magical in training. Technically, training is doing a series of
+**mathematical computation** (requiring these op kernels), and update the weight
+**variables** which is **accessible by other (inference) subgraphs**. The
+converter also needs to be changed, to support and optimize training use cases.
+
+### Sharing variables across subgraphs
+
+For the variable requirement of TensorFlow Lite training, the high level goals
+are:
+
+*   Convert TensorFlow Resource Variables to TensorFlow Lite, and preserve the
+    same semantic
+*   Able to share variables between multiple TensorFlow Lite subgraphs
+
+We propose to define variables in TFLite that have similar semantics to resource
+variables in TensorFlow:
+
+*   Use int32 tensor to represent variable ID in TFLite instead of defining a
+    resource type
+*   Only 2 ops are required to make it work: `AssignVariableOp` and
+    `ReadVariableOp`
+*   Other ops can be added into TFLite when necessary
+
+### Gradient and optimizer operations
+
+To train a model, it's required to compute **gradients**, and apply the gradient
+to trainable variables with **optimizers** (e.g. SGD, ADAM...etc). Technically,
+gradients and optimizers are just some mathematical computation and variable
+read/write operations. When constructing training graphs in TensorFlow,
+sometimes a single fused op is used (e.g. `ReluGrad`), and sometimes it produces
+multiple regular ops in unfused form (e.g. `Mul`, `MatMul`, `Add`) to compute
+the gradient of one op.
+
+Today TensorFlow Lite doesn't have these training-specific ops. The following
+approaches are considered to run training ops in TensorFlow Lite:
+
+**Add training ops into Flex runtime (initial implementation)**<br/>
+
+Note: "Flex" is the code name for
+[Select TensorFlow operators to use in TensorFlow Lite](https://www.tensorflow.org/lite/guide/ops_select)
+project, which enables using TensorFlow kernels in TensorFlow Lite directly.
+Throughout the document, it will be referred as "Flex" to be concise.
+
+This is the easiest approach: Just whitelist these training ops to Flex whitelist. However this means Flex runtime is required for training (bloating binary size), and Flex kernels are not optimized for mobile.
+
+**Implement fused ops as TensorFlow Lite builtin ops (optimization)**<br/>
+We can implement these fused ops as TensorFlow Lite builtin ops. This further
+enables optimizing for mobile devices (e.g. writing optimized SIMD CPU kernels,
+writing GPU delegate...etc). This may need a significant amount of work.
```

Comment:
How about the support of the Pixel Visual core during training?
It seems that currently, Android Neural Network only has inference APIs.

Reply:
The software architecture should allow delegating the training computation to hardwares (though the TensorFlow Lite [Delegate API](https://www.tensorflow.org/lite/performance/delegates)). So far we didn't look into specific delegate yet (e.g. NNAPI, GPU) but it should be doable.